libgadget/lenstools.c

```diff
@@ -0,0 +1,412 @@
+#include "lenstools.h"
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <fftw3.h> 
+
+#ifdef USE_CFITSIO
+#include "fitsio.h"
+#endif
+
+#include <string.h>
+#include "partmanager.h"
+#include "cosmology.h"
+#include "physconst.h"
+
+void linspace(double start, double stop, int num, double *result) {
+    double step = (stop - start) / (num - 1);
+    for (int i = 0; i < num; i++) {
+        result[i] = start + i * step;
+    }
+}
+
+double ***allocate_3d_array(int nx, int ny, int nz) {
+    // Allocate memory for the 3D array
+    double ***array = malloc(nx * sizeof(double **));
+    if (array == NULL) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(EXIT_FAILURE);
+    }
+
+    for (int i = 0; i < nx; i++) {
+        array[i] = malloc(ny * sizeof(double *));
+        if (array[i] == NULL) {
+            fprintf(stderr, "Memory allocation failed for layer %d\n", i);
+            exit(EXIT_FAILURE);
+        }
+        for (int j = 0; j < ny; j++) {
+            array[i][j] = malloc(nz * sizeof(double));
+            if (array[i][j] == NULL) {
+                fprintf(stderr, "Memory allocation failed for layer %d, row %d\n", i, j);
+                exit(EXIT_FAILURE);
+            }
+            // Initialize elements (optional)
+            for (int k = 0; k < nz; k++) {
+                array[i][j][k] = 0.0; // Example initialization
+            }
+        }
+    }
+
+    return array;
+}
+
+void free_3d_array(double ***array, int nx, int ny) {
+    // Free each row
+    for (int i = 0; i < nx; i++) {
+        for (int j = 0; j < ny; j++) {
+            free(array[i][j]);
+        }
+        free(array[i]);
+    }
+    free(array);
+}
+
+
+// Function to allocate a 2D array
+double **allocate_2d_array(int Nx, int Ny) {
+    double **array = malloc(Nx * sizeof(double *));
+    if (array == NULL) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(EXIT_FAILURE);
+    }
+    for (int i = 0; i < Nx; i++) {
+        array[i] = malloc(Ny * sizeof(double));
+    }
+    return array;
+}
+
+// Function to free a 2D array
+void free_2d_array(double **array, int Nx) {
+    for (int i = 0; i < Nx; i++) {
+        free(array[i]);
+    }
+    free(array);
+}
+
+typedef struct {
+    int nx, ny, nz;
+} GridDimensions;
+
+// Function to determine the bin index for a given value
+int find_bin(double value, double *bins, int num_bins) {
+    for (int i = 0; i < num_bins - 1; i++) {
+        if (value >= bins[i] && value < bins[i + 1]) {
+            return i;
+        }
+    }
+    return -1;  // Return the last bin if the value is on the boundary or beyond
+}
+
+void grid3d_nfw(const struct particle_data * Parts, int num_particles, double **binning, GridDimensions dims, double ***density) {
+    
+    double position[3];
+    // Process each particle
+    for (int p = 0; p < num_particles; p++) {
```

Comment:
We want this to be #pragma omp parallel for but then density[] will be written to from different threads and will race. In petapm the similar code guards density with #pragma omp atomic, but it might be more effective to allocate separate arrays for each  processor if density is small.

Reply:
If I remember, we do not want to make it omp parallel anymore?

Reply:
No want to make it omp parallel but not MPI parallel